processing/src/main/java/io/druid/query/PrioritizedExecutorService.java

```diff
@@ -185,89 +200,130 @@ public int getQueueSize()
   {
     return delegateQueue.size();
   }
+}
 
-
-  public static class PrioritizedListenableFutureTask<V> implements RunnableFuture<V>,
-      ListenableFuture<V>,
-      PrioritizedRunnable,
-      Comparable<PrioritizedListenableFutureTask>
+class PrioritizedListenableFutureTask<V> implements RunnableFuture<V>,
+    ListenableFuture<V>,
+    PrioritizedRunnable,
+    Comparable<PrioritizedListenableFutureTask>
+{
+  private static final AtomicLong insertionDownCounter = new AtomicLong(Long.MAX_VALUE);
+  // NOTE: For these LOWER means higher priority. As such we swap left and right in the compares
+  private static final Comparator<PrioritizedListenableFutureTask> PRIORITY_COMPARATOR = new Ordering<PrioritizedListenableFutureTask>()
   {
-    public static <V> PrioritizedListenableFutureTask<V> create(PrioritizedRunnable task, @Nullable V result)
+    @Override
+    public int compare(
+        PrioritizedListenableFutureTask left, PrioritizedListenableFutureTask right
+    )
     {
-      return new PrioritizedListenableFutureTask<>(ListenableFutureTask.create(task, result), task.getPriority());
+      return Integer.compare(right.getPriority(), left.getPriority());
     }
-
-    public static <V> PrioritizedListenableFutureTask<?> create(PrioritizedCallable<V> callable)
+  }.nullsLast();
+  private static final Comparator<PrioritizedListenableFutureTask> PRIORITY_AND_ENUMERATION_COMPARATOR = new Ordering<PrioritizedListenableFutureTask>()
+  {
+    @Override
+    public int compare(PrioritizedListenableFutureTask left, PrioritizedListenableFutureTask right)
     {
-      return new PrioritizedListenableFutureTask<>(ListenableFutureTask.create(callable), callable.getPriority());
+      int compare = PRIORITY_COMPARATOR.compare(left, right);
+      if (compare == 0) {
+        compare = Long.compare(right.getInsertionPlace(), left.getInsertionPlace());
+      }
+      return compare;
     }
+  }.nullsLast();
 
-    public static <V> PrioritizedListenableFutureTask<V> create(ListenableFutureTask<V> task, int priority)
-    {
-      return new PrioritizedListenableFutureTask<>(task, priority);
-    }
+  public static <V> PrioritizedListenableFutureTask<V> create(
+      PrioritizedRunnable task,
+      @Nullable V result,
+      boolean fifo
```

Comment:
is seems odd to have FIFO be a property of the item in the queue, I feel this makes more sense to be a property of the executorservice itself. What if instead of passing the fifo flag, we give each task a sequence number. The executorservice can then decide whether to pass 0 all the time or to pass actual sequence numbers, and it also simplifies the compareTo, because we always use the same comparator regardless of being fifo or not.